File: Go/Fibonacci.go
package main

import "fmt"

func main() {
  fmt.Println(fibo(80) == 23416728348467685)
}


func mul(u [2][2]int, v [2][2]int) [2][2]int {
  a, b := u[0][0], u[0][1]
  c, d := u[1][0], u[1][1]
  e, f := v[0][0], v[0][1]
  g, h := v[1][0], v[1][1]
  return [2][2]int {{a*e+b*g, a*f+b*h}, {c*e+d*g, c*f+d*h}}
}

func exp(u [2][2]int, n int) [2][2]int {
  if n == 0 { return [2][2]int {{1, 0}, {0, 1}} }
  if n == 1 { return u }
  return mul(exp(u, n/2), exp(u, (n + n%2)/2))
}

func fibo(n int) int {
  a := [2][2] int {{1,1},{1,0}}
  res := exp(a, n)
  return res[0][1]
}

File: Rust/Fibonacci.rs
type L = (u64, u64, u64, u64);

fn main() {
  let idx = 50;
  println!("{}th Fibonacci number is {}!", idx, fibo(idx));
}

fn fibo(n: u16) -> u64 {
  let x = (1, 1, 1, 0);
  let x = pow(x, n);
  return x.1
}

fn pow(m: L, n: u16) -> L {
  if n == 0 { return (1, 0, 0, 1) }
  if n == 1 { return m }
  return mul(pow(m, n/2), pow(m, n/2+n%2))
}

fn mul(m: L, n: L) -> L {
  let (a, b, c, d) = m;
  let (e, f, g, h) = n;
  return (a*e+b*g, a*f+b*h, c*e+d*g, c*f+d*h)
}
